### ISSUE TITLE
incorrect outout for bitfield load using i64

### ISSUE THREAD

**Issue by sunfishcode** [WebAssembly/binaryen#407]
The following testcase, compiled with the llvm-wasm backend, runs with incorrect output under binaryen. It appears to be related to i64.

```
#include <stdio.h>

struct S1 { unsigned lo:32; unsigned mid:2; unsigned hi:30; };

static struct S1 g_68 = { -1, 3, 0xbbe };

int main() {
    printf("%x\n", g_68.hi);
    return 0;
}

```

**Comment by kripken** [comment]
Where is that from, btw?

**Comment by sunfishcode** [comment]
It's reduced from tests/fuzz/17.c

**Comment by kripken** [comment]
Looking into this, I removed the printf and replaced it with direct logging calls, to not involve libc,

````
#include <stdint.h>
#include <stdio.h>

struct S1 { unsigned lo:32; unsigned mid:2; unsigned hi:30; };

static struct S1 g_68 = { -1, 0, 0xbbe };

extern "C" void emscripten_autodebug_i32(int32_t x, int32_t y);

int main() {
  emscripten_autodebug_i32(0, g_68.lo);
  emscripten_autodebug_i32(1, g_68.mid);
  emscripten_autodebug_i32(2, g_68.hi);
  return 0;
}````

and that now prints out the 3 elements in that struct. The code the wasm backend emitted looks weird to me,

````
    (call_import $emscripten_autodebug_i32
      (get_local $$0)
      (i32.load offset=1032
        (get_local $$0)
      )
    )
    (call_import $emscripten_autodebug_i32
      (i32.const 1)
      (i32.wrap/i64
        (i64.and
          (i64.load32_u offset=1036
            (get_local $$0)
          )
          (i64.const 3)
        )
      )
    )
    (call_import $emscripten_autodebug_i32
      (i32.const 2)
      (i32.wrap/i64
        (i64.shr_u
          (i64.load offset=1032 align=4
            (get_local $$0)
          )
          (i64.const 34)
        )
      )
    )
````

And the data segment is

````
    (segment 1032 "\ff\ff\ff\ff\f8.\00\00")
````

which doesn't look right. That is -1 for the first 32 bits, correctly, but then 0xf82e0000 for the other 32 bits, which doesn't look like the 0x00000bbe I would expect? The .s is the same,

````
_ZL4g_68:
        .int8   255                     # 0xff
        .int8   255                     # 0xff
        .int8   255                     # 0xff
        .int8   255                     # 0xff
        .int8   251                     # 0xfb
        .int8   46                      # 0x2e
        .int8   0                       # 0x0
        .int8   0                       # 0x0
        .size   _ZL4g_68, 8
````

**Comment by binji** [comment]
Looks ok to me, `0xbbe` in 30-bits is:
```
                          b    b    e
00 0000 0000 0000 0000 1011 1011 1110
```

so the full 64-bits:
```
                         0xbbe  0                               -1
000000000000000000101110111110 00 11111111111111111111111111111111

00000000 00000000 00101110 11111000 11111111 11111111 11111111 11111111
      00       00       2e       f8       ff       ff       ff       ff
```

The strange part is that `0xfb` in the .s file. It should be `0xf8`.
Oh, I see. The original example was `{ -1, 3, 0xbbe }` and you changed it `{ -1, 0, 0xbbe }`

**Comment by kripken** [comment]
Yes, I simplified the middle part to 0.

Oh right, it's shifted by 2. Then that part does look right...

**Comment by kripken** [comment]
Fix in #460.
